File: android/app/src/main/java/my/ssdid/mobile/domain/model/DidDocument.kt
package my.ssdid.mobile.domain.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class DidDocument(
    @SerialName("@context") val context: List<String> = listOf("https://www.w3.org/ns/did/v1"),
    val id: String,
    val controller: String,
    val verificationMethod: List<VerificationMethod>,
    val authentication: List<String>,
    val assertionMethod: List<String>,
    val capabilityInvocation: List<String> = emptyList()
) {
    companion object {
        fun build(did: Did, keyId: String, algorithm: Algorithm, publicKeyMultibase: String): DidDocument {
            return DidDocument(
                id = did.value,
                controller = did.value,
                verificationMethod = listOf(
                    VerificationMethod(
                        id = keyId,
                        type = algorithm.w3cType,
                        controller = did.value,
                        publicKeyMultibase = publicKeyMultibase
                    )
                ),
                authentication = listOf(keyId),
                assertionMethod = listOf(keyId),
                capabilityInvocation = listOf(keyId)
            )
        }
    }
}

@Serializable
data class VerificationMethod(
    val id: String,
    val type: String,
    val controller: String,
    val publicKeyMultibase: String
)

File: sdk/ios/SsdidCore/Sources/SsdidCore/Domain/Model/DidDocument.swift
import Foundation

public struct DidDocument: Codable, Sendable {
    public let context: [String]
    public let id: String
    public var controller: String = ""
    public let verificationMethod: [VerificationMethod]
    public var authentication: [String] = []
    public var assertionMethod: [String] = []
    public var capabilityInvocation: [String] = []
    public var nextKeyHash: String? = nil

    enum CodingKeys: String, CodingKey {
        case context = "@context"
        case id, controller, verificationMethod
        case authentication, assertionMethod, capabilityInvocation, nextKeyHash
    }

    public init(
        context: [String] = ["https://www.w3.org/ns/did/v1"],
        id: String,
        controller: String = "",
        verificationMethod: [VerificationMethod],
        authentication: [String] = [],
        assertionMethod: [String] = [],
        capabilityInvocation: [String] = [],
        nextKeyHash: String? = nil
    ) {
        self.context = context
        self.id = id
        self.controller = controller
        self.verificationMethod = verificationMethod
        self.authentication = authentication
        self.assertionMethod = assertionMethod
        self.capabilityInvocation = capabilityInvocation
        self.nextKeyHash = nextKeyHash
    }

    public static func build(
        did: Did,
        keyId: String,
        algorithm: Algorithm,
        publicKeyMultibase: String
    ) -> DidDocument {
        DidDocument(
            id: did.value,
            controller: did.value,
            verificationMethod: [
                VerificationMethod(
                    id: keyId,
                    type: algorithm.w3cType,
                    controller: did.value,
                    publicKeyMultibase: publicKeyMultibase
                )
            ],
            authentication: [keyId],
            assertionMethod: [keyId],
            capabilityInvocation: [keyId]
        )
    }
}
